src/averageAges.js

```diff
@@ -1,61 +1,61 @@
 'use strict';
 
-/**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = people.filter(person =>
+    century
+      ? person.sex === 'm' && Math.ceil(person.died / 100) === century
+      : person.sex === 'm'
+  );
+  const sumAge = men.reduce(
+    (sum, person) => sum + (person.died - person.born),
+    0
+  );
+
+  return sumAge / men.length;
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people.filter(person => person.sex === 'f');
+  const womenWithChildren = withChildren
+    ? women.filter(woman => people.some(person => person.mother === woman.name))
+    : women;
+
+  const sumAge = womenWithChildren.reduce(
+    (sum, person) => sum + (person.died - person.born),
+    0
+  );
+
+  return sumAge / womenWithChildren.length;
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothers = onlyWithSon
+    ? people.filter(woman =>
+      people.some(
+        person => person.mother === woman.name && person.sex === 'm'))
+    : people.filter(woman =>
+      people.some(person => person.mother === woman.name));
```

Comment:
Refactor it to filter people only once. Move a ternary operator inside the filter method

Reply:
`  const mothers = people.filter(woman =>
    onlyWithSon
      ? people.some(person =>
        person.mother === woman.name && person.sex === 'm')
      : people.some(person => person.mother === woman.name)
  );`

I changed the logic, filter it once, and then check if condition withSon is true, if yes then we perform first version of some method, if not we just look for all children.

Further nesting some statement would add an unnecesarry level of looping for checking onlyWithSons condition, to add i find it more readable.